Pathogens/Utility/location_utils/location_functions.py

```diff
@@ -178,21 +178,51 @@ def compute_pin_info(df: pd.DataFrame, geo_dfs: dict) -> pd.DataFrame:
 
     return df
 
-def get_lng_lat(geocoder_search_text, geocoder_data_type, country_code=None):
-    # If a city doesn't have a state associated with it,
-    # we cannot accurately find its location
-    if (not geocoder_search_text) or (geocoder_data_type == 'place' and "," not in geocoder_search_text):
+def get_mapbox_api_query_url(geocoder_search_text, geocoder_data_type, country_code):
+    return f"https://api.mapbox.com/geocoding/v5/mapbox.places/{geocoder_search_text}.json?" \
+          f"access_token={os.getenv('MAPBOX_API_KEY')}&types={geocoder_data_type}&country={country_code}"
+
+def parse_mapbox_response(response):
+    data = response.json()
+    if data and "features" in data and len(data['features']) > 0:
+        return data['features'][0]['center']
+    else:
         return None
 
-    url = f"https://api.mapbox.com/geocoding/v5/mapbox.places/{geocoder_search_text}.json?" \
-          f"access_token={os.getenv('MAPBOX_API_KEY')}&types={geocoder_data_type}"
+def get_city_lat_lng(city_name, state_name, country_name):
+    if(city_name is None):
+        return get_state_lat_lng(state_name, country_name)
 
-    # if country_code:
-    #     url += f"&country={country_code}"
+    mapbox_search_text = city_name + ',' + state_name if (state_name is not None) else city_name
+    
+    url = get_mapbox_api_query_url(mapbox_search_text, 'place', country_code=get_country_code(country_name=country_name, iso3=False))
+
+    api_response = requests.get(url)
+    coords = parse_mapbox_response(api_response)
+
+    if(coords is None):
+        return get_state_lat_lng(state_name, country_name)
+    else:
+        return coords
+
+def get_state_lat_lng(state_name, country_name):
+    if(state_name is None):
+        return get_country_lat_lng(country_name)
+
+    url = get_mapbox_api_query_url(state_name, 'region', country_code=get_country_code(country_name=country_name, iso3=False))
+
+    api_response = requests.get(url)
+    coords = parse_mapbox_response(api_response)
+
+    if(coords is None):
+        return get_country_lat_lng(country_name)
+    else:
+        return coords
+
+def get_country_lat_lng(country_name):
+    url = get_mapbox_api_query_url(country_name, 'country', country_code=get_country_code(country_name=country_name, iso3=False))
+
+    api_response = requests.get(url)
+    coords = parse_mapbox_response(api_response)
 
```

Comment:
I think one of the things we probably want to consider here is if mapbox allows batch requests and if we can do that so instead of having 1102 requests we have at max 3.